catalog/src/main/java/it/eng/catalog/service/ArtifactService.java

```diff
@@ -44,45 +44,56 @@ public ArtifactService(MongoTemplate mongoTemplate, DatasetService datasetServic
 
 	public List<Artifact> getArtifacts(String artifactId) {
 		if(StringUtils.isNotBlank(artifactId)) {
-			return artifactRepository.findById(artifactId)
-					.stream()
-					.collect(Collectors.toList());
+			Optional<Artifact> artifact = artifactRepository.findById(artifactId);
+			if (artifact.isPresent()) {
+				return List.of(artifact.get());
+			} else {
+				throw new ResourceNotFoundAPIException("Artifact with id " + artifactId + " not found");
+			}
 		}
 		return artifactRepository.findAll();
 	}
 
-	public Artifact uploadArtifact(MultipartFile file, String datasetId, URL externalURL) {
+	public Artifact uploadArtifact(MultipartFile file, String datasetId, String externalURL) {
+		Dataset dataset = datasetService.getDatasetByIdForApi(datasetId);
 		if (file != null) {
-			return storeFile(file, datasetId);
+			return storeFile(file, dataset);
 		} else if (externalURL != null) {
-			return insertExternalArtifact(datasetId, externalURL);
+			return insertExternalArtifact(dataset, externalURL);
 		}
 		log.warn("Artifact and file not found");
 		throw new CatalogErrorAPIException("Artifact and file not found");
 	}
+	
+	public void deleteArtifact (String id) {
+		Artifact artifact = artifactRepository.findById(id)
+				.orElseThrow(() -> new CatalogErrorAPIException("Could not delete artifact, artifact with id: " + id + " does not exist"));
+		deleteArtifactAndData(artifact);
```

Comment:
Should we remove from dataset first, and if this is OK, then delete Artifact/file from Mongo?
Also good candidate for Transaction